File: integrations/sqlite/typescript/scripts/benchmark-writes.mjs
// Run after building Core and SQLite: node scripts/benchmark-writes.mjs
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { InMemoryRunRepository, assertRunRepositoryConforms } from "purra";
import { SqliteAgentAdapters } from "../dist/index.js";

const reference = new InMemoryRunRepository();
await assertRunRepositoryConforms(reference);
const { preset, budgets } = await reference.get("conformance-run-1");
for (const count of [100, 1000, 5000]) {
  const dir = mkdtempSync(join(tmpdir(), "purra-write-bench-"));
  const storage = new SqliteAgentAdapters(join(dir, "agent.db"), { scope: "benchmark" });
  try {
    await storage.transaction(async ({ runs }) => {
      await runs.begin({ preset, budgets, requestedRunId: "benchmark", deadlineAt: null, metadata: {} });
      for (let index = 0; index < count; index++) {
        await runs.appendEvent("benchmark", { sourceKey: `event:${index}`, kind: "model.diagnostics",
          channel: "model", visibility: "private", payload: { text: "x".repeat(128) } });
      }
    });
    const elapsed = [];
    for (let attempt = 0; attempt < 12; attempt++) {
      const start = performance.now();
      const result = await storage.idempotency.executeOnce(`tool:${attempt}`, async () => ({ content: "committed" }));
      elapsed.push(performance.now() - start);
      if (result.content !== "committed") throw new Error("missing receipt");
    }
    const sorted = elapsed.slice(2).sort((a, b) => a - b);
    console.log(JSON.stringify({ history_events: count, median_receipt_write_ms: Number(((sorted[4] + sorted[5]) / 2).toFixed(3)), measured_writes: 10 }));
  } finally { storage.close(); rmSync(dir, { recursive: true }); }
}

File: integrations/sqlite/python/scripts/benchmark_writes.py
"""Measure tool receipt writes with a growing canonical output journal.

Run from the repository root:
PYTHONPATH=src:integrations/sqlite/python/src .venv/bin/python integrations/sqlite/python/scripts/benchmark_writes.py
"""

import asyncio
from datetime import datetime, timezone
import json
from pathlib import Path
import statistics
import tempfile
import time

from purra.contracts import RunCreateParams, ToolCall, ToolHandlerResult
from purra.events import AgentEvent
from purra.output.contracts import AgentOutputEventDraft
from purra_sqlite import SqliteAgentAdapters


async def main():
    for count in (100, 1000, 5000):
        with tempfile.TemporaryDirectory() as directory:
            storage = SqliteAgentAdapters(Path(directory) / "agent.db", scope="benchmark")
            try:
                async with storage.transaction() as adapters:
                    run = (await adapters.runs.begin(RunCreateParams(None, "benchmark", None), AgentEvent("run.started"))).run_id
                    for index in range(count):
                        await adapters.outputs.append_event(AgentOutputEventDraft(
                            run_id=run, turn_id=None, output_stream_id=None, invocation_id=None,
                            source_event_key=f"event:{index}", source="domain", kind="domain.effect",
                            channel="diagnostic", visibility="private", payload={"text": "x" * 128},
                            occurred_at=datetime.now(timezone.utc),
                        ))
                async def effect():
                    return ToolHandlerResult("committed")
                elapsed = []
                for attempt in range(12):
                    start = time.perf_counter()
                    result = await storage.idempotency.execute_once(run, ToolCall(f"tool:{attempt}", "write", "{}"), effect)
                    elapsed.append((time.perf_counter() - start) * 1000)
                    assert result.content == "committed"
                print(json.dumps({"history_events": count, "median_receipt_write_ms": round(statistics.median(elapsed[2:]), 3), "measured_writes": 10}), flush=True)
            finally:
                storage.close()


if __name__ == "__main__":
    asyncio.run(main())
